### PR TITLE
Improve isImage() check

### PR DESCRIPTION
**Pull request by caneleex**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [x] Documentation
- [ ] Other: \_____ 

Closes Issue: #1071

## Description

Improve isImage() check by using image pattern.

### CODE REVIEW

**Comment by JohnnyJayJay** [comment]
I would make that pattern a constant of the `Message` interface or at least not make it a local variable, since it never changes. That way it wouldn't need to be compiled for every call and users may utilise it, too.

**Comment by natanbc** [comment]
You can make it `static final`, there's no need to create a pattern for each instance.

**Comment by Devoxin** [comment]
This will not work (reliably) if the attachment has had its file extension changed.



A way around this would be to check `width > -1` (or height, I guess. JDA returns -1 if the attachment is not an image, assuming JDA pulls this [information straight from the Message object itself](https://discordapp.com/developers/docs/resources/channel#attachment-object-attachment-structure)), along with a file extension check (or if you're feeling super confident, parse the image contents and detect file type).

Lastly, there are more image types out there than what the regex supports, unless this is only intended to support images that Discord can embed.

File: src/main/java/net/dv8tion/jda/api/entities/Message.java [hunk 1]
```diff
@@ -1117,6 +1117,13 @@ public Pattern getPattern()
         private final int size;
         private final int height;
         private final int width;
+        
+        /**
+        * Pattern used to check if the file is an image.
+        *
+        * @see #isImage()
+        */
+        private final Pattern image = Pattern.compile("([^\\s]+)(?:\\.(?i)(jpe?g|png|gif|bmp|webp))$");
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/Message.java]: Why use a regex for this? This is trivial to implement without a regex and would probably be less error-prone than this.

```java
int index = name.lastIndexOf('.');
if (index + 1 == name.length())
    return false;
String ext = name.substring(index + 1);
return "png|webp|jpeg|jpg|gif".contains(ext);
```

Also I'm pretty sure bmp is not supported.

Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/entities/Message.java]: You're right probably. I'll leave the decision up to you and other maintainers.

**Comment by jagrosh** [comment]
As already stated, a regex would be a really slow solution to this problem. Also, the width and height should still be checked (in addition to whatever is used to separate images from video) to ensure that the image is actually being shown.

**Comment by DV8FromTheWorld** [comment]
1. We should consider adding `Attachment#getFileExtension()` and use that in this PR
2. `Attachment#isImage()`should _first_ check if there are width / height values before doing the additional extension check.
3. The extension check should not use Regex.

**Comment by caneleex** [comment]
Why would you all firstly check for width and height and not just the extension check? Isn't it kinda useless?

**Comment by Devoxin** [comment]
Because as per the discord docs, width and height properties exist for attachments that are images. Checking those properties first would be a "less expensive" operation and then the extension check would serve as an extra step to ensure it's an image and not a video (although, I'm not sure if Discord sends width/height properties for videos so even the extension check might be useless here?)

**Comment by JohnnyJayJay** [comment]
The extension check is not useless because extensions don't determine the file format, just how it's displayed by Discord.
I.e. if you do not check the extension, someone could upload a raw text file with a `.jpg` extension and this method would return true.
From what I understand, even with the extension check it's not 100% safe yet, since someone still could upload a video with an image extension. But it seems like the reasonably closest we can get.

**Comment by Devoxin** [comment]
You can send a txt file disguised as a jpg/png/whatever and Discord will not send width/height properties for it (they will be null, or in JDA's case, `-1`), so this is not a problem.

However, I have just checked to see if the width/height properties exist for videos, and they do. 

So the width/height check just checks if the attachment is a video/image, and the extension check would serve to filter out videos. Though, it is worth mentioning that **videos disguised as images by changing the extension, will still have width/height properties**. End result is that this proposed check will still return a false positive in this case.

**Comment by caneleex** [comment]
This is really confusing and I have no idea of reliable check.

**Comment by DV8FromTheWorld** [comment]
At the end of the day, we can only provide as much security as we are given. We can ensure it is at least something that is visually displayable via the width and height checks, and we can _attempt_ to provider a further restriction by using file extension checking, but at the end of the day, if it really really matters

**Comment by MinnDevelopment** [comment]
@caneleex looks like you deleted your branch, I assume you don't intend to continue working on this PR then? In that case could you close it so someone else can give it a shot.

Thank you for your contributions.

**Comment by caneleex** [comment]
I think that this PR doesn't solve the reliability of the check and someone could've done it better; so yeah, I will close the PR and let someone other try to solve it. I also think this isn't very easy to fix as obviously there is no 100% reliable fix.

**Pull request closed without merging**